### PR TITLE
slqx migration, credential processing and cookie authentication

### PR DESCRIPTION
**Pull request by realaravinth**
This PR is massive and I apologize for that. The `sqlx` migration work touched various areas so I figured I'd refactor those areas as well.

## Changes:
- `sqlx` migration
- credential processing with [argon2-creds](https://github.com/realaravinth/argon2-creds)
- `jwt` to cookie-based authentication based
- tests for authentication module
- CI now runs integration tests against a live Postgres database
- `actix-governor` is disabled. I don't know how to enable it also I found that a new middleware instance was created for each service so does that mean that each service gets its own counter or is there a shared counter? Please feel free to make necessary changes.

### CODE REVIEW

**Comment by AaronErhardt** [comment]
Wow, that's huge! Thank you so much for your work!
I have an exam tomorrow so I don't have time today, but I will review these patches after the exam :)

**Comment by realaravinth** [comment]
sure, take your time! and best of luck for the test! :)

File: src/apps/files/upload.rs [hunk 1]
```diff
@@ -18,41 +16,40 @@ struct Response {
...
-            let mut file = fs::File::create(file_path).await?;
-
-            // Field in turn is stream of *Bytes* object
-            while let Some(chunk) = field.next().await {
-                file.write_all(&chunk?).await?;
+        if let Some(content_type) = field.content_disposition() {
+            if let Some(filename) = content_type.get_filename() {
+                if filename.contains("..") {
+                    return Err(ServiceError::PermissionDenied);
+                }
```
Comment (AaronErhardt) [inline src/apps/files/upload.rs:37]: I guess this is another place where cap-std could be useful...

File: src/tests/helpers.rs [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+#[macro_export]
+macro_rules! post_request {
+    ($uri:expr) => {
+        test::TestRequest::post().uri($uri)
+    };
+
+    ($serializable:expr, $uri:expr) => {
+        test::TestRequest::post()
+            .uri($uri)
+            //.insert_header((header::CONTENT_TYPE, "application/json"))
```
Comment (AaronErhardt) [inline src/tests/helpers.rs:46]: I guess the commented line here can be removed

**Review by AaronErhardt — CHANGES_REQUESTED** [review]
That was quite a lot to review but looks pretty good overall! There were just a few minor things that need to be adjusted.

**Comment by AaronErhardt** [comment]
Also I just noticed that if no Postgres server is running, the connection timeout is pretty high which causes the program to run for quite some time before realizing that there's no server to connect to. Also in this phase while waiting for the database pool (which should be pretty much idle) the server has a pretty high CPU usage for some reason...

File: src/apps/files/mod.rs [hunk 3]
```diff
@@ -38,28 +38,28 @@ pub struct SourceAndDest {
...
 /// Helper function to translate paths from requests into absolute path
-fn resolve_path(user_id: u32, query_path: &str) -> ServiceResult<std::path::PathBuf> {
+fn resolve_path(username: &str, query_path: &str) -> ServiceResult<std::path::PathBuf> {
     if query_path.contains("..") {
         Err(ServiceError::PermissionDenied)
     } else {
         Ok(std::path::PathBuf::from(format!(
             "data/users/{}/files/{}",
-            user_id, query_path
+            username, query_path
```
[the remark this answers is not in the document]
Comment (AaronErhardt) [reply src/apps/files/mod.rs:47]: I just checked and it seems like this isn't possible but the server responds with "username_case_mapped violation". So it seems that argon2-creds with case mapping solves this issue. Still cap-std could improve the robustness of the code.

File: src/app_state.rs [hunk 4]
```diff
@@ -1,30 +1,43 @@
...
+        let c = creds.clone();
 
-        // create database pool
-        let db_pool = database::connect(&config.database.url())
-            .expect("Failed to create database pool.");
+        let init = std::thread::spawn(move || {
+            log::info!("Initializing credential manager");
+            c.init();
+            log::info!("Initialized credential manager");
+        });
```
[the remark this answers is not in the document]
Comment (realaravinth) [reply src/app_state.rs:33]: `argon2-creds` uses lazily initialized regex filters, they are large and take time to initialize. By calling `init`, we are initializing at startup where we can afford a little delay.  

And no, it isn't lost. Regex filters are `static` and `arong2-creds` has handles to them at all times.

File: src/api/v1/auth.rs [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    Ok(HttpResponse::Ok())
+}
+
+#[my_codegen::get(path = "crate::V1_API_ROUTES.auth.logout")] // wrap = "crate::CheckLogin")]
+async fn signout(id: Identity) -> impl Responder {
+    if id.identity().is_some() {
+        id.forget();
+    }
+    HttpResponse::Found()
+        //        .append_header((header::LOCATION, "/login"))
```
[the remark this answers is not in the document]
Comment (realaravinth) [reply src/api/v1/auth.rs:226]: That's `actix-web` v4 stuff. We will be migrating to v4 soon

File: Cargo.toml [hunk 6]
```diff
@@ -44,7 +45,14 @@ actix-files = "0.5"
...
 actix-web = { version = "3", features = ["openssl"] }
-openssl = "0.10"
+actix-identity = "0.3"
+actix-service = "1.0.6"
+openssl = { version = "0.10.29", features = ["vendored"] }
+
+sqlx = { version = "0.4.0", features = [ "runtime-actix-rustls", "postgres", "time", "offline" ] }
+argon2-creds = { branch = "master", git = "https://github.com/realaravinth/argon2-creds"}
+lazy_static = "1.4"
+my-codegen = {package = "actix-web-codegen", git ="https://github.com/realaravinth/actix-web"}
```
[the remark this answers is not in the document]
Comment (realaravinth) [reply Cargo.toml:55]: Two reasons:
- main actox-web-codegen doesn't take `expr` for route. IMO, having `expr` for route is nice as you can save route in a `const` and use it everywhere (tests, templates, etc)
- this is more of an after thought: I'm thinking about moving `src/tests/helper.rs` into the same crate.

File: src/api/v1/account/delete.rs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+    let rec = sqlx::query_as!(
+        Password,
+        r#"SELECT password  FROM triox_users WHERE name = ($1)"#,
+        &username,
+    )
+    .fetch_one(&data.db)
+    .await;
+
+    id.forget();
```
[the remark this answers is not in the document]
Comment (realaravinth) [reply src/api/v1/account/delete.rs:47]: Ah, good catch!

File: src/api/v1/auth.rs [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+        };
+
+        if payload.login.contains('@') {
+            #[derive(Clone, Debug)]
+            struct EmailLogin {
+                name: String,
+                password: String,
+            }
+
+            let email_fut = sqlx::query_as!(
```
[the remark this answers is not in the document]
Comment (realaravinth) [reply src/api/v1/auth.rs:94]: Good catch! It's baggage from previous iterations, I'll be sure to clean it up.

File: src/api/v1/auth.rs [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        }
+    }
+
+    pub async fn register_runner(
+        payload: &Register,
+        data: &AppData,
+    ) -> ServiceResult<()> {
+        //  if !crate::SETTINGS.server.allow_registration {
+        //      return Err(ServiceError::ClosedForRegistration);
+        //  }
```
[the remark this answers is not in the document]
Comment (realaravinth) [reply src/api/v1/auth.rs:137]: I was going to make a PR for the same :)

File: src/apps/files/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (realaravinth) [reply src/apps/files/mod.rs:47]: You are right about argon2-creds`, it does username case mapped validation. 

`cap-std` looks cool but that is for another PR and it requires serious thought. We should also look into implement some access control model

File: src/apps/files/upload.rs (hunk 1 above)
Comment (realaravinth) [reply src/apps/files/upload.rs:37]: [Please see previous comment](https://github.com/AaronErhardt/Triox/pull/34#discussion_r665864354)

File: src/tests/helpers.rs (hunk 2 above)
Comment (realaravinth) [reply src/tests/helpers.rs:46]: That's actix-web v4 stuff

**Comment by realaravinth** [comment]
IMO, the timeout is alright as it comes in handy when theres a start up delay in situations like systemd service set to launch thus program at boot time or launching a trail net with docker compose. 

The high CPU usage is caused by `argon2-creds` initializing regex filters. In release builds, it should be brief but in debug, it's awfully long.

File: Cargo.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (AaronErhardt) [reply Cargo.toml:55]: Makes sense. Do plan to upstream these features?

[the remark this answers is not in the document]
Comment (realaravinth) [reply Cargo.toml:55]: I did but there wasn't enough response to warrant inclusion into the main lib. See [#2216](https://github.com/actix/actix-web/pull/2216). The fork will be maintained and publish to crates.io.

File: src/apps/files/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (AaronErhardt) [reply src/apps/files/mod.rs:47]: I agree there needs to be a more sophisticated method for handling files. Btw. I still plan on implementing some sort of encrypted federated block storage that can easily be synced. I already started with a FUSE file system implementation but it's still far from finished. Maybe I will find more time for it in a few weeks...

**Comment by realaravinth** [comment]
I've added a coverage workflow. Tarpaulin is failing on my computer with same error(broken pipe) but it seems to function well in CI runs on my other repositories. We'll see if I can get it running on here.

File: src/apps/files/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (realaravinth) [reply src/apps/files/mod.rs:47]: I've been thinking about a block based implementation as well. We'll approach things step by step, this way it'll give us room to experiment and pick methods that work best for us. For now(as in this PR), this seems adequate. 

I want to resolve access control issues with web UI before implementing files backend and its access control mechanisms. Currently, unauthorized access is possible for `/static/files.html`. It won't list files any users but it's possible to view the webpage.

**Review by AaronErhardt — APPROVED** [review]
Looks good to me! I think we can merge the PR now.

**Pull request merged**